Android 12 Firebase dynamic link was not well handled

AdrienAudouard opened the pull request:
The deeplink created by firebase was not well handled for Android 12 and above. 

A firebase deeplink is composed by a short link (who should be received by the app) and a link (this link is used when the user do not have the app installed on his device).
Before Android 12 the short link was send to the app but from Android 12 and above the link was send to the app, so it break the deeplink behavior on some apps. This PR fix this and always send the correct link.

llfbandit commented on app_links/android/src/main/java/com/llfbandit/app_links/AppLinksHelper.java:
```diff
@@ -0,0 +1,102 @@
+    }
+
+    public int validateObjectHeader(@NonNull Parcel p) {
+        int var1 = p.readInt();
```
Please, name correctly the fields:
var1 => header
var2 => size
var 3=> start
20293 => MAGIC_NUMBER

llfbandit commented on app_links/android/src/main/java/com/llfbandit/app_links/AppLinksHelper.java:
```diff
@@ -0,0 +1,102 @@
+        int var2 = readSize(p, var1);
+        int var3 = p.dataPosition();
+        if (getFieldId(var1) != 20293) {
+            throw new NullPointerException();
```
IllegalFormatException instead.

llfbandit commented on app_links/android/src/main/java/com/llfbandit/app_links/AppLinksHelper.java:
```diff
@@ -0,0 +1,102 @@
+        int var3 = p.dataPosition();
+        if (getFieldId(var1) != 20293) {
+            throw new NullPointerException();
+        } else {
```
Remove else

llfbandit commented on app_links/android/src/main/java/com/llfbandit/app_links/AppLinksHelper.java:
```diff
@@ -0,0 +1,102 @@
+            throw new NullPointerException();
+        } else {
+            var1 = var3 + var2;
+            if (var1 >= var3 && var1 <= p.dataSize()) {
```
Rework this block:
```java
int end = start + size;
if ((end < start) || (end > parcel.dataSize())) {
  throw new IllegalFormatException();
}
return end;
```

llfbandit commented on app_links/android/src/main/java/com/llfbandit/app_links/AppLinksHelper.java:
```diff
@@ -0,0 +1,102 @@
+        int var2 = p.dataPosition();
+        if (header == 0) {
+            return null;
+        } else {
```
Please, name variables correctly.
Remove useless else keyword.

llfbandit commented on app_links/example/windows/flutter/generated_plugins.cmake:
```diff
@@ -6,6 +6,9 @@ list(APPEND FLUTTER_PLUGIN_LIST
   app_links_windows
 )
 
+list(APPEND FLUTTER_FFI_PLUGIN_LIST
```
Why Windows cmake is impacted?

llfbandit commented on app_links/android/src/main/java/com/llfbandit/app_links/AppLinksPlugin.java:
```diff
@@ -170,11 +171,9 @@ private boolean handleIntent(Intent intent) {
 
-    String action = intent.getAction();
-    String dataString = intent.getDataString();
+    AppLinksHelper helper = new AppLinksHelper();
```
Methods on AppLinksHelper should be static:
- to improve performance
- to not ambiguate its responsability.

llfbandit reviewed (changes requested):
Thanks for the proposal.
Appart the requested tiny changes, seems good to me.
Also, some additional comments could be great to know where this source code comes from.

AdrienAudouard commented:
Hey, thanks for your response.
So, I removed all the code related to the data reading and I've added the `org.microg:safe-parcel` dependancy which is doing the same job (my code was roughly inspired by this lib).

llfbandit commented on app_links/android/src/main/java/com/llfbandit/app_links/AppLinksPlugin.java:
```diff
@@ -2,6 +2,7 @@
 
 import android.app.Activity;
 import android.content.Intent;
+import android.os.Parcel;
```
Unused import?

llfbandit commented on app_links/android/src/main/java/com/llfbandit/app_links/AppLinksHelper.java:
```diff
@@ -0,0 +1,63 @@
+        while (parcel.dataPosition() < end) {
+            int header = parcel.readInt();
+            switch (SafeParcelReader.getFieldId(header)) {
+                case 1:
```
What are cases 1 and 2 and why iterate if there's only one single read in the parcel?

llfbandit commented on app_links/android/src/main/java/com/llfbandit/app_links/AppLinksHelper.java:
```diff
@@ -0,0 +1,63 @@
+
+    public static String getDeepLinkFromIntent(Intent intent) {
+        String shortLink = getShortDeepLink(intent);
+        Log.d(TAG, "handleIntent: (Action) " + intent.getExtras().toString());
```
Duplicated logs between this place and `getUrl` method.

llfbandit reviewed (changes requested):
Almost there! Thanks for the quick changes.

llfbandit commented:
> Hey, thanks for your response. So, I removed all the code related to the data reading and I've added the `org.microg:safe-parcel` dependancy which is doing the same job (my code was roughly inspired by this lib).

I know :)
You code was fine without the dependancy but it's ok like this too. Unused code should be trimmed at build time.

AdrienAudouard commented:
I'm not really of copying the code of another lib but if you prefer I can rollback this change !
